File: bin/send_to_honeycomb.py
#!/usr/bin/env python3
"""
Send workflow metrics.jsonl to Honeycomb.

Usage:
    python3 bin/send_to_honeycomb.py \\
      --honeycomb-key YOUR_API_KEY \\
      --dataset workflow-metrics \\
      /path/to/workflow/state

    # Optional environment
    python3 bin/send_to_honeycomb.py \\
      --honeycomb-key YOUR_API_KEY \\
      --dataset workflow-metrics \\
      --env us \\  # or 'eu'
      /path/to/workflow/state
"""

import json
import sys
import argparse
from pathlib import Path
from typing import List, Dict, Any
import urllib.request
import urllib.error

class HoneycombClient:
    """Simple Honeycomb API client."""

    def __init__(self, api_key: str, dataset: str, env: str = "us"):
        self.api_key = api_key
        self.dataset = dataset
        self.env = env
        self.endpoint = f"https://api.honeycomb.io/1/events/{dataset}"
        self.sent = 0
        self.failed = 0

    def send_event(self, event: Dict[str, Any]) -> bool:
        """Send a single event to Honeycomb."""
        try:
            data = json.dumps(event).encode("utf-8")
            req = urllib.request.Request(
                self.endpoint,
                data=data,
                headers={
                    "X-Honeycomb-Team": self.api_key,
                    "Content-Type": "application/json",
                },
            )

            with urllib.request.urlopen(req, timeout=10) as response:
                if response.status == 200:
                    self.sent += 1
                    return True
                else:
                    print(f"❌ HTTP {response.status}: {response.read().decode()}")
                    self.failed += 1
                    return False

        except urllib.error.HTTPError as e:
            print(f"❌ HTTP Error {e.code}: {e.read().decode()}")
            self.failed += 1
            return False
        except Exception as e:
            print(f"❌ Error: {e}")
            self.failed += 1
            return False

    def send_batch(self, events: List[Dict[str, Any]]) -> int:
        """Send multiple events."""
        success = 0
        for event in events:
            if self.send_event(event):
                success += 1
        return success


def load_metrics(state_root: Path) -> List[Dict[str, Any]]:
    """Load all metrics.jsonl files from issues directories."""
    metrics = []
    for metrics_file in state_root.glob("issues/*/metrics.jsonl"):
        with open(metrics_file) as f:
            for line in f:
                if line.strip():
                    try:
                        metrics.append(json.loads(line))
                    except json.JSONDecodeError as e:
                        print(f"⚠️  Invalid JSON in {metrics_file}: {e}")
    return metrics


def enrich_events(events: List[Dict[str, Any]], repo: str = None) -> List[Dict[str, Any]]:
    """Enrich events with additional fields for Honeycomb."""
    enriched = []

    for event in events:
        # Calculate derived fields for easier querying
        if event.get("event") == "close_completed":
            dispositions = event.get("dispositions", [])
            total = len(dispositions)
            if total > 0:
                addressed = sum(1 for d in dispositions if d["disposition"] == "addressed")
                event["human_addressed_pct"] = round((addressed / total) * 100, 1)
            else:
                event["human_addressed_pct"] = 0

        if event.get("event") == "review_completed":
            total_findings = (
                event.get("critical_count", 0)
                + event.get("minor_count", 0)
                + event.get("notes_count", 0)
            )
            event["findings_total"] = total_findings

        if repo:
            event["repository"] = repo

        enriched.append(event)

    return enriched


def main():
    parser = argparse.ArgumentParser(
        description="Send workflow metrics to Honeycomb"
    )
    parser.add_argument(
        "--honeycomb-key",
        required=True,
        help="Honeycomb API key",
    )
    parser.add_argument(
        "--dataset",
        required=True,
        help="Honeycomb dataset name",
    )
    parser.add_argument(
        "--env",
        default="us",
        choices=["us", "eu"],
        help="Honeycomb environment",
    )
    parser.add_argument(
        "--repo",
        help="Repository name to tag events",
    )
    parser.add_argument(
        "--batch-size",
        type=int,
        default=50,
        help="Events per batch (default: 50)",
    )
    parser.add_argument(
        "state_root",
        help="Path to workflow/state branch root",
    )

    args = parser.parse_args()

    state_root = Path(args.state_root)
    if not state_root.exists():
        print(f"❌ Path not found: {state_root}")
        sys.exit(1)

    print(f"📊 Loading metrics from {state_root}...")
    metrics = load_metrics(state_root)

    if not metrics:
        print("❌ No metrics found")
        sys.exit(1)

    print(f"✅ Loaded {len(metrics)} events")

    print(f"🔧 Enriching events...")
    enriched = enrich_events(metrics, repo=args.repo)

    print(f"📤 Sending to Honeycomb ({args.env})...")
    print(f"   Dataset: {args.dataset}")
    print(f"   Batch size: {args.batch_size}")

    client = HoneycombClient(args.honeycomb_key, args.dataset, args.env)

    # Send in batches for rate limiting
    for i in range(0, len(enriched), args.batch_size):
        batch = enriched[i : i + args.batch_size]
        print(f"   Batch {i//args.batch_size + 1}: ", end="", flush=True)
        success = client.send_batch(batch)
        print(f"{success}/{len(batch)} ✓")

    print()
    print("=" * 50)
    print(f"📈 Results:")
    print(f"   Sent:   {client.sent}")
    print(f"   Failed: {client.failed}")

    if client.failed > 0:
        print(f"\n⚠️  Some events failed to send. Check:")
        print(f"   - API key is valid")
        print(f"   - Dataset exists in Honeycomb")
        print(f"   - Network connectivity")
        sys.exit(1)
    else:
        print(f"\n✨ All events sent successfully!")
        print(f"\nNext steps:")
        print(f"  1. Go to honeycomb.io")
        print(f"  2. Open dataset: {args.dataset}")
        print(f"  3. Build queries and dashboards!")

if __name__ == "__main__":
    main()

File: bin/send-to-honeycomb.js
#!/usr/bin/env node
/**
 * Send workflow metrics.jsonl to Honeycomb
 *
 * Usage:
 *   node bin/send-to-honeycomb.js \
 *     --honeycomb-key YOUR_API_KEY \
 *     --dataset workflow-metrics \
 *     /path/to/workflow/state
 */

const fs = require("fs");
const path = require("path");
const https = require("https");

class HoneycombClient {
  constructor(apiKey, dataset, env = "us") {
    this.apiKey = apiKey;
    this.dataset = dataset;
    this.env = env;
    this.endpoint = `api.honeycomb.io`;
    this.sent = 0;
    this.failed = 0;
  }

  sendEvent(event) {
    return new Promise((resolve) => {
      const data = JSON.stringify(event);
      const options = {
        hostname: this.endpoint,
        port: 443,
        path: `/1/events/${this.dataset}`,
        method: "POST",
        headers: {
          "X-Honeycomb-Team": this.apiKey,
          "Content-Type": "application/json",
          "Content-Length": Buffer.byteLength(data),
        },
        timeout: 10000,
      };

      const req = https.request(options, (res) => {
        if (res.statusCode === 200) {
          this.sent++;
          resolve(true);
        } else {
          console.error(`❌ HTTP ${res.statusCode}`);
          this.failed++;
          resolve(false);
        }
        res.on("data", () => {}); // consume response
      });

      req.on("error", (e) => {
        console.error(`❌ Error: ${e.message}`);
        this.failed++;
        resolve(false);
      });

      req.on("timeout", () => {
        req.destroy();
        this.failed++;
        resolve(false);
      });

      req.write(data);
      req.end();
    });
  }

  async sendBatch(events) {
    let success = 0;
    for (const event of events) {
      if (await this.sendEvent(event)) {
        success++;
      }
    }
    return success;
  }
}

function loadMetrics(stateRoot) {
  const metrics = [];
  const issuesDir = path.join(stateRoot, "issues");

  if (!fs.existsSync(issuesDir)) {
    return metrics;
  }

  const issues = fs.readdirSync(issuesDir);
  for (const issue of issues) {
    const metricsFile = path.join(issuesDir, issue, "metrics.jsonl");
    if (fs.existsSync(metricsFile)) {
      const lines = fs
        .readFileSync(metricsFile, "utf-8")
        .split("\n")
        .filter((line) => line.trim());

      for (const line of lines) {
        try {
          metrics.push(JSON.parse(line));
        } catch (e) {
          console.error(`⚠️  Invalid JSON in ${metricsFile}: ${e.message}`);
        }
      }
    }
  }

  return metrics;
}

function enrichEvents(events, repo) {
  return events.map((event) => {
    // Add derived fields for Honeycomb
    if (event.event === "close_completed") {
      const dispositions = event.dispositions || [];
      const total = dispositions.length;
      if (total > 0) {
        const addressed = dispositions.filter(
          (d) => d.disposition === "addressed"
        ).length;
        event.human_addressed_pct = Math.round((addressed / total) * 100 * 10) / 10;
      } else {
        event.human_addressed_pct = 0;
      }
    }

    if (event.event === "review_completed") {
      const total =
        (event.critical_count || 0) +
        (event.minor_count || 0) +
        (event.notes_count || 0);
      event.findings_total = total;
    }

    if (repo) {
      event.repository = repo;
    }

    return event;
  });
}

async function main() {
  const args = process.argv.slice(2);

  let apiKey;
  let dataset;
  let env = "us";
  let repo;
  let batchSize = 50;
  let stateRoot;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--honeycomb-key") {
      apiKey = args[++i];
    } else if (args[i] === "--dataset") {
      dataset = args[++i];
    } else if (args[i] === "--env") {
      env = args[++i];
    } else if (args[i] === "--repo") {
      repo = args[++i];
    } else if (args[i] === "--batch-size") {
      batchSize = parseInt(args[++i]);
    } else if (!args[i].startsWith("--")) {
      stateRoot = args[i];
    }
  }

  if (!apiKey || !dataset || !stateRoot) {
    console.error("Usage:");
    console.error(
      "  node send-to-honeycomb.js --honeycomb-key KEY --dataset DATASET [--repo NAME] /path/to/workflow/state"
    );
    process.exit(1);
  }

  if (!fs.existsSync(stateRoot)) {
    console.error(`❌ Path not found: ${stateRoot}`);
    process.exit(1);
  }

  console.log(`📊 Loading metrics from ${stateRoot}...`);
  const metrics = loadMetrics(stateRoot);

  if (metrics.length === 0) {
    console.log("ℹ️  No metrics found");
    process.exit(0);
  }

  console.log(`✅ Loaded ${metrics.length} events`);
  console.log(`🔧 Enriching events...`);

  const enriched = enrichEvents(metrics, repo);

  console.log(`📤 Sending to Honeycomb (${env})...`);
  console.log(`   Dataset: ${dataset}`);
  console.log(`   Batch size: ${batchSize}`);

  const client = new HoneycombClient(apiKey, dataset, env);

  for (let i = 0; i < enriched.length; i += batchSize) {
    const batch = enriched.slice(i, i + batchSize);
    process.stdout.write(
      `   Batch ${Math.floor(i / batchSize) + 1}: `,
      "utf-8"
    );
    const success = await client.sendBatch(batch);
    console.log(`${success}/${batch.length} ✓`);
  }

  console.log("");
  console.log("=".repeat(50));
  console.log(`📈 Results:`);
  console.log(`   Sent:   ${client.sent}`);
  console.log(`   Failed: ${client.failed}`);

  if (client.failed > 0) {
    console.error(`\n⚠️  Some events failed to send. Check:`);
    console.error(`   - API key is valid`);
    console.error(`   - Dataset exists in Honeycomb`);
    console.error(`   - Network connectivity`);
    process.exit(1);
  } else {
    console.log(`\n✨ All events sent successfully!`);
  }
}

main().catch((e) => {
  console.error(`Fatal error: ${e.message}`);
  process.exit(1);
});
